Added checksum tracking and enhanced trim workloads

phoulihan9 opened the pull request.

paulhoulihan committed:
Added checksum tracking and enhanced trim workloads

sitsofe commented:
Hi, this seems to be a pull request from your repo against your own repo but I can't tell if this was the intent? If you meant to submit the pull request to Jens you can do something like the following:

1. Find your branch in Github (e.g. go to https://github.com/phoulihan9/fio because in your case the master branch has your change in it)
2. Click on New Pull Request .
3. This should take you to https://github.com/axboe/fio/compare/master...phoulihan9:master where you will submit a request to the axboe repo master branch to pull the changes from the phoulihan9 master branch.

The one thing I will say is this is a big patch. It might be helpful to split it into several smaller commits because as you've got general fixes, additional workloads, additional tools AND the new checksum verification all in the same commit.

sitsofe commented on analyze/corruption_triage.sh:
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
```
Perhaps it would be better for the analyze directory to live under tools/analyze ?

sitsofe commented on backend.c:
```diff
@@ -1428,9 +1441,10 @@ static uint64_t do_dry_run(struct thread_data *td)
+		// TODO Setting verify_backlog could result in verify reads from get_io_u
+		// TODO that would not be performed when we fake completion and thus not all writes
+		// TODO would be verified in a dry run.
+		if (verifiable_io(td, io_u) &&
```
It would be useful to bail out with an explanation at startup time if your constraints aren't met e.g. in `fixup_options()` of init.c .

sitsofe commented on debug.h:
```diff
@@ -11,6 +11,7 @@ enum {
 	FD_BLKTRACE,
 	FD_VERIFY,
+	FD_CHKSUM,
 	FD_RANDOM,
```
It might be safer to add this new option just before the end of the enum ...

sitsofe commented on engines/windowsaio.c:
```diff
@@ -182,8 +182,7 @@ static int fio_windowsaio_open_file(struct thread_data *td, struct fio_file *f)
 	 */
-	if (td->o.td_ddir == TD_DDIR_READ  ||
-		td->o.td_ddir == TD_DDIR_WRITE)
+	if (!td_random(td))
```
(Not introduced by this commit but an idle comment) Technically shouldn't this only be done if the user has set [`fadvise_hint`](http://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-fadvise-hint) ?

sitsofe commented on fio.1:
```diff
@@ -344,27 +344,60 @@ Sequential reads.
+ - Often minimal block size required. Linux on VMware requires 1 MB
+ - VMware requires minimum VM OS hardware level 11
+ - To verify the trim I/Os requires verify_track
+.RE
```
I don't know you can say much more about the requirements other than "you may need extra privileges and your OS/Storage may enforce a minimum trim block size along with other constraints" e.g. I thought with ESXi 6.0 you had to also set EnableBlockDelete on the VM for UNMAP commands to work (obviously this has changed with ESXi 6.5)?

sitsofe commented on io_ddir.h:
```diff
@@ -27,26 +33,51 @@ static inline const char *io_ddir_name(enum fio_ddir ddir)
+#define td_randrwt(td)			((td)->o.td_ddir == TD_DDIR_RANDRWT)
+
+// Return mask of I/O types
+#define td_eligible_ddirs(td)   ((td)->o.td_ddir & TD_DDIR_RWT)
```
What are these ddirs eligible for - doing I/O that sends new data to a block?

sitsofe commented on io_u.c:
```diff
@@ -661,9 +729,9 @@ int io_u_quiesce(struct thread_data *td)
-	long usec;
-	uint64_t now;
+	enum fio_ddir min_odir;
+	long usec, now, min_time;
```
Experience says keeping now in "only" a long will lead to unsigned integer overflow on some platforms: https://github.com/axboe/fio/commit/90eff1c9f01d2f8b4ff8bc75a2bf120a1789b37c .

sitsofe commented on options.c:
```diff
@@ -1608,19 +1643,47 @@ struct fio_option fio_options[FIO_MAX_OPTS] = {
 			  { .ival = "trimwrite",
 			    .oval = TD_DDIR_TRIMWRITE,
-			    .help = "Trim and write mix, trims preceding writes"
+				.help = "Sequential trim then write, trim preceding write on each block"
```
Indentation looks different to preceding options...

sitsofe commented on time.c:
```diff
@@ -63,8 +63,17 @@ uint64_t usec_sleep(struct thread_data *td, unsigned long usec)
+		}
+		// nanosleep was hanging on Windows 7 so replaced
+		//if (nanosleep(&req, NULL) < 0)
+		//	break;
```
Rather than just stubbing this out we need to know why it went wrong. @phoulihan9 do you have the workload that generated the problem? Perhaps @bcran has some suggestions on why this happened?

sitsofe commented:
As you're adding a new option / things that will impact client/server fields you will have to bump `FIO_SERVER_VER` in server.h up by one too. In fact - does client/server mode work with these changes?

My comments are those of a random passer-by so I cherry picked areas and entirely skipped over others. Hopefully they're helpful...

phoulihan9 commented:
Great comments Sitsofe, I respond to each below.

However can you or Jens give me an idea on what the review process is like
and when will I have all the review comments in? How long will this take?


>Hi, this seems to be a pull request from your repo against your own repo
but I can't tell if this was the intent? If you meant to submit the pull
request to

>Jens you can do something like the following:

   1. Find your branch in Github (e.g. go to https://github.com/
   phoulihan9/fio because in your case the master branch has your change in
   it)
   2. Click on New Pull Request .
   3. This should take you to axboe@master...phoulihan9:master
   <https://github.com/axboe/fio/compare/master...phoulihan9:master> where
   you will submit a request to the axboe repo master branch to pull the
   changes from the phoulihan9 master branch.

>The one thing I will say is this is a big patch. It might be helpful to
split it into several smaller commits because as you've got general fixes,
additional

>workloads, additional tools AND the new checksum verification all in the
same commit.

I will fix the branch.

Separate commits could be a lot of work. Regretfully its all in one place.
I especially wanted the new rw= workloads separated from the new verify
functionality. But the full solution required that they be somewhat
intertwined. Ideally this is good advice but if creating separate commits
is a requirement then this is going to take a long time to finally be
completed.

-------------------------------

>In analyze/corruption_triage.sh
<https://github.com/phoulihan9/fio/pull/1#discussion_r115410511>:

>> @@ -0,0 +1,143 @@
>+#!/bin/bash

>Perhaps it would be better for the analyze directory to live under
tools/analyze ?
I agree that's a better place for these tools.

----------------------------

>+		// TODO Setting verify_backlog could result in verify reads from get_io_u
>+		// TODO that would not be performed when we fake completion and thus not all writes
>+		// TODO would be verified in a dry run.
>+		if (verifiable_io(td, io_u) &&

>It would be useful to bail out with an explanation at startup time if your
constraints aren't met e.g. in fixup_options() of init.c .
Easy enough for me to do if you buy my argument. I have not tested out the
theory but seems obvious enough. If you agree with me then I will add a
restrictions to init.c.

---------------------------------------

>+	FD_CHKSUM,
> 	FD_RANDOM,

>It might be safer to add this new option just before the end of the enum
...

Yes, of course, what was I thinking?

---------------------------------------------------


>-	if (td->o.td_ddir == TD_DDIR_READ  ||
>-		td->o.td_ddir == TD_DDIR_WRITE)
>+	if (!td_random(td))

>(Not introduced by this commit but an idle comment) Technically shouldn't
this only be done if the user has set fadvise_hint
<http://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-fadvise-hint>
 ?

Seems reasonable But I don't know this code well enough. I can put a TODO
in the code so the suggestion is not lost. What was there didn't look right
and it was important to me to get all the references to o.td_ddir to be
through the defined macros.

--------------------------------


>+ - File system and OS support varies but Linux block devices accept trims.
>+   You need privilege to write to a Linux block device.
>+   See example fio: track-mem.fio
>+ - Often minimal block size required. Linux on VMware requires 1 MB
>+ - VMware requires minimum VM OS hardware level 11
>+ - To verify the trim I/Os requires verify_track
>+.RE

>I don't know you can say much more about the requirements other than "you
may need extra privileges and your OS/Storage may enforce a minimum trim

>block size along with other constraints" e.g. I thought with ESXi 6.0 you
had to also set EnableBlockDelete on the VM for UNMAP commands to work

>(obviously this has changed with ESXi 6.5)?

I would like to keep the specific details so that people will test with
trims. These restrictions are easy to work around if you know what they are
but to the causal user, they are insurmountable without the details.
Possibly I missed the EnableDeleteBlock as that was placed on by default in
my template. I will check this.

--------------------------------

>+// Return mask of I/O types
>+#define td_eligible_ddirs(td)   ((td)->o.td_ddir & TD_DDIR_RWT)

>What are these ddirs eligible for - doing I/O that sends new data to a
block?

Yes these are the bits in the ddir mask that involve some form of I/O,
eligible for consideration when asking about the type of I/O. Would you
prefer a better name?

-----------------------------------

>+	enum fio_ddir min_odir;
>+	long usec, now, min_time;

>Experience says keeping now in "only" a long will lead to unsigned integer
overflow on some platforms: axboe@90eff1c
<https://github.com/axboe/fio/commit/90eff1c9f01d2f8b4ff8bc75a2bf120a1789b37c>
 .

I think I changed this as there was some compiler issue I was responding
to. I can look into it.

---------------------------------------------

>-			    .help = "Trim and write mix, trims preceding writes"
>+				.help = "Sequential trim then write, trim preceding write on each block"

>Indentation looks different to preceding options...

Fixed.

--------------------------------------------

>+		// nanosleep was hanging on Windows 7 so replaced
>+		//if (nanosleep(&req, NULL) < 0)
>+		//	break;

>Rather than just stubbing this out we need to know why it went wrong.
@phoulihan9 <https://github.com/phoulihan9> do you have the workload that
generated the problem?

>Perhaps @bcran <https://github.com/bcran> has some suggestions on why this
happened?
I assume it was buried in the guts of cygwin and didn't have time to
pursue. This is a blocker on all versions of windows and so its important.
Its a big help to QA & dev to have tools and scripts that can run on linux
and windows without change.

----------------------------------------------

> As you're adding a new option / things that will impact client/server
fields you will have to bump FIO_SERVER_VER in server.h up by one too. In
fact -

>does client/server mode work with these changes?

>My comments are those of a random passer-by so I cherry picked areas and
entirely skipped over others. Hopefully they're helpful...

Have not tried client/server. This change is easy enough but not sure I
know enough about client server to give it a good test.



On Tue, May 9, 2017 at 1:38 AM, Sitsofe Wheeler <notifications@github.com>
wrote:

> Hi, this seems to be a pull request from your repo against your own repo
> but I can't tell if this was the intent? If you meant to submit the pull
> request to Jens you can do something like the following:
>
>    1. Find your branch in Github (e.g. go to
>    https://github.com/phoulihan9/fio <https://github.com/phoulihan9/fio>
>    because in your case the master branch has your change in it)
>    2. Click on New Pull Request .
>    3. This should take you to axboe@master...phoulihan9:master
>    <https://github.com/axboe/fio/compare/master...phoulihan9:master>
>    where you will submit a request to the axboe repo master branch to pull the
>    changes from the phoulihan9 master branch.
>
> The one thing I will say is this is a big patch. It might be helpful to
> split it into several smaller commits because as you've got general fixes,
> additional workloads, additional tools AND the new checksum verification
> all in the same commit.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/phoulihan9/fio/pull/1#issuecomment-300066285>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AalS2Jk6_K9gNyk_-DB5WrOBmSttrwISks5r3_vRgaJpZM4NTYRy>
> .
>

sitsofe commented:
Mentioning @axboe because he probably hasn't seen these comments.

> However can you or Jens give me an idea on what the review process is like
and when will I have all the review comments in? How long will this take?

Ultimately you won't need my review if you convince Jens it's mergeable ;-) - I'm just trying to be helpful. I can't give a definite time for when I finish review as I do this in my spare time which is limited and sporadic and I'm volunteering. I also can't tell you "you only have to do X and Y and then it it will be guaranteed merged" which is harsh but it's the best I can do. I know Jens goes through periods of busyness which means he may not get round to seeing an updated pull request for several weeks...

> Separate commits could be a lot of work. Regretfully its all in one place.
I especially wanted the new rw= workloads separated from the new verify
functionality. But the full solution required that they be somewhat
intertwined. Ideally this is good advice but if creating separate commits
is a requirement then this is going to take a long time to finally be
completed.

I definitely agree separate commits are extra work but I think they'll lead to a speedier merge. My thinking is:

- Some of the changes look "easy" to accept immediately.
- The core checksumming feature might need more evolving before being merged because it has a lot of constraints.
- There are "controversial" changes (such as the commenting out of the Windows' nanosleep  and changes to default fio behaviour) which I think would be blockers to the whole commit as it stands but wouldn't stop the prior pieces going in by themselves.
- If this work were merged and some major design issue is found it would be nice to just be able to revert just pieces of it.

> > >
 ```patch
-	if (td->o.td_ddir == TD_DDIR_READ  ||
-		td->o.td_ddir == TD_DDIR_WRITE)
+	if (!td_random(td))
```
> > (Not introduced by this commit but an idle comment) Technically shouldn't
this only be done if the user has set fadvise_hint
> > <http://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-fadvise-hint>
 ?
>
> Seems reasonable But I don't know this code well enough. I can put a TODO
in the code so the suggestion is not lost. What was there didn't look right
and it was important to me to get all the references to o.td_ddir to be
through the defined macros.

We might be able to ask others about this on the mailing list.

```
+// Return mask of I/O types
+#define td_eligible_ddirs(td)   ((td)->o.td_ddir & TD_DDIR_RWT)
```
> > What are these ddirs eligible for - doing I/O that sends new data to a
block?
>
> Yes these are the bits in the ddir mask that involve some form of I/O,
> eligible for consideration when asking about the type of I/O. Would you
> prefer a better name?

For consideration for checksumming/tracking for verify? If so perhaps "verifiable" would be a nicer name...

```patch
+		// nanosleep was hanging on Windows 7 so replaced
+		//if (nanosleep(&req, NULL) < 0)
+		//	break;
```

> > Rather than just stubbing this out we need to know why it went wrong.
> > @phoulihan9 <https://github.com/phoulihan9> do you have the workload that
> > generated the problem?
> > Perhaps @bcran <https://github.com/bcran> has some suggestions on why this
happened?
>
> I assume it was buried in the guts of cygwin and didn't have time to
> pursue. This is a blocker on all versions of windows and so its important.
> Its a big help to QA & dev to have tools and scripts that can run on linux
> and windows without change.

Understandable but do you know if there's "plain fio-2.19" job file that also shows the issue? It might be possible to deal with this outside of this commit...

axboe commented:
Sitsofe is right that separate commits is the way to go, for several reasons:

1) It's much easier to review. One big patch with tons of different changes in it is almost impossible to review properly.
2) If unintended changes or bugs occur as a side effect of your changes, then having smaller patches allows you to much more easily figure out where things broke, and fix that up. git bisect is a wonderful tool for that.

Bigger changes should always be implemented as a series of smaller changes.

What I've done in the past with a big patch is follow this cycle:

1) Edit out stand-alone feature or prep change from the big patch
2) Apply and commit above stand alone change
3) Kill that change from the big patch, which now applies on top

Repeat until each patch only does one logical thing. You will probably still end up with a bigger feature patch for your series, but at least you will have split out some of the changes following that procedure.

Your patch is especially problematic, since you are changing behavior for and are potentially adding features that overlap. This is much easier to see and modify, if each change is much smaller. I have to feel confident applying a patch. If I don't, then I will not do it. This might mean that fio misses out on a useful piece of functionality, but long term maintainability and stability of fio is more important to me.

I'd love to review and apply your changes, but I just cannot pull in a massive change like that. It absolutely has to be split up into digestible pieces first.

phoulihan9 commented:
>Your patch is especially problematic, since you are changing behavior for
and are potentially adding features that overlap. This is much easier to
see and

>modify, if each change is much smaller. I have to feel confident applying
a patch. If I don't, then I will not do it. This might mean that fio misses
out on a

>useful piece of functionality, but long term maintainability and stability
of fio is more important to me.
OK, good that tells me where I stand and I understand your priorities.
Unfortunately I am taking a new job and need to concentrate on it so have
run out of time. Perhaps in the future I can break these changes apart once
I get set up on new hardware and can find the time.

It would be good to know if its worth my time though, especially for the
changes to basic workloads. Here are the major issues I see (and please
mention any non-starter issues you or Sitsofe see):

1) Sitsofe mentioned that are quite a number of restrictions with this new
tracking log. My team never used these restricted features in data
integrity testing and I think they are quite esoteric. With effort,
potentially some restrictions could be lifted in the future. Can you review
the list of restrictions (see verify_track description in HOWTO) and let me
know if the list is viable as is?

2) I changed the rw=readwrite workload to make a full pass of the file
using a single file pointer rather than separate last_pos pointers for
reads and writes. This makes more sense to me and tends to be what people
expected as derived from the other sequential workloads. See my initial
description bullet starting with "rw=readwrite seems to have independent " for
more detail. Would you be willing to make this change or does old behavior
have to be maintained?

3) I changed how bytes are counted towards the TOTAL_BYTES as workload
counting rules were inconsistent. Now read verify bytes are not being
counted towards workload totals and this can affect the number of I/Os
performed in a workload but makes fio more predictable. See my initial
description bullet starting with "Counting read verification bytes" for
more detail. Are you willing to accept this change?

If these issues are nonstarters then maybe I should consider creating a fio
spinoff focused on data integrity and just go with what I have. Maybe call
it "fio-tracking" or something. Of course there is plenty of downside with
this strategy. Is there any precedent for this approach?

paul

On Fri, May 12, 2017 at 11:56 AM, Jens Axboe <notifications@github.com>
wrote:

> Sitsofe is right that separate commits is the way to go, for several
> reasons:
>
>    1. It's much easier to review. One big patch with tons of different
>    changes in it is almost impossible to review properly.
>    2. If unintended changes or bugs occur as a side effect of your
>    changes, then having smaller patches allows you to much more easily figure
>    out where things broke, and fix that up. git bisect is a wonderful tool for
>    that.
>
> Bigger changes should always be implemented as a series of smaller changes.
>
> What I've done in the past with a big patch is follow this cycle:
>
>    1. Edit out stand-alone feature or prep change from the big patch
>    2. Apply and commit above stand alone change
>    3. Kill that change from the big patch, which now applies on top
>
> Repeat until each patch only does one logical thing. You will probably
> still end up with a bigger feature patch for your series, but at least you
> will have split out some of the changes following that procedure.
>
> Your patch is especially problematic, since you are changing behavior for
> and are potentially adding features that overlap. This is much easier to
> see and modify, if each change is much smaller. I have to feel confident
> applying a patch. If I don't, then I will not do it. This might mean that
> fio misses out on a useful piece of functionality, but long term
> maintainability and stability of fio is more important to me.
>
> I'd love to review and apply your changes, but I just cannot pull in a
> massive change like that. It absolutely has to be split up into digestible
> pieces first.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/phoulihan9/fio/pull/1#issuecomment-301115832>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AalS2MW3vspL8kQdTQwr46oE11eXD-8Hks5r5IE6gaJpZM4NTYRy>
> .
>

sitsofe commented:
Bear in mind there's a chicken and an the egg issue here: because the commit is so big it takes review to find out what you don't like about it but it's hard to review because it's so big etc.

Non-starters you didn't mention for me you didn't directly mention (i.e. things other than changes in defaults) in the current commit:
- `now` variable being changed to a long.
- nanosleep change which introduces a change but just comments out the old code. Really needs a small reproducer to show the problem.

2) Sitsofe mentioned that are quite a number of restrictions with this new tracking log.

   The restrictions that Paul refers to are mostly mentioned in https://github.com/phoulihan9/fio/pull/1/files/7aaeea5a483e59d356432a469bd897298a495192#diff-05bc0511c8db43874bb452c70f7b6ddfR2626 . For convenience I'll reproduce them here    
   ```patch
   +    back on verification described under the verify option. The restrictions when
   +    enabling the verify_track option are:
   +            - randommap is required
   +            - softrandommap is not supported
   +            - lfsr random generator not supported when using multiple block sizes
   +            - stonewall option required when more than one job present
   +            - file size must be an even multiple of the block size when iodepth > 1
   +            - verify_backlog not supported when iodepth > 1
   +            - verify_async is not supported
   +            - file sharing between concurrent jobs not supported
   +            - numjobs must be 1
   +            - io_submit_mode must be set to "inline"
   +            - verify=null or pattern are not supported
   +            - verify_only is not supported
   +            - io_submit_mode must be set to 'inline'
   +            - supplying a sequence number with rw option is not supported
   +            - experimental_verify is not supported
   ```
 
   The ones that stand out are:
   - stonewall required (even if the files aren't shared?)
   - verify_backlog not supported when iodepth > 1
   - numjobs being 1 (even if the files aren't shared?),
   - verify_only is not supported (doesn't that make multiple re-verifications harder?)

   (PS: io_submit_mode seems to be listed twice)

   It's worth noting that there are definitely fio users using bsrange with verification because they report corner cases where it causes spurious mismatches. Personally, I've recently been using fio (with patches!) to do non-uniform verified overlapping I/O (in my case by using a random_distribution).

Re fio-tracking: forks happen and I'd vastly prefer a fork than this work being lost/unpublished :-) You wouldn't have to go so far as naming your fork entirely differently but perhaps tweak the VERSION string so it's clear you're not mainstream fio. However, there's a lot in this commit that I think can go into upstream fio so if a fork were to occur perhaps it would be possible to periodically submit small pieces of the fork upstream and thus shrinking it over time...

(PS: @phoulihan9  your replies look a bit weird on github's web interface. You might find using github's web interface useful if you need to make a more complicated/intricate reply)

phoulihan9 commented:
Sitsofe,

I didn't mention your other issues as I think they are workable but the
issues I mentioned would amount to non-starters for big hunks of the
checkin. Those chunks would not be worth trying to port if I don't get some
conceptual buyin beforehand.

I would like to see fio parameters resulting in overlapping I/O due to
bsrange. My research led me to believe bsrange alone was safe and was not
added to the list of restrictions associated verify_track. Overlaps more
likely resulted from readwrite workload using separate last_pos pointers
rather than bsrange alone.

So I would like to create a fork with a version of "2.19-tracking" and
offer it for availability. Then as time permits start moving the changes
into fio mainstream. A few things can easily be separate checkins. The big
cost is that the new workloads will be hard to slice off from the tracking
log. To separate them would require development and test of new
transitional code/state that would be discarded in the long run, which
poses its own level of risk given that its not in regular use by a test
organization on multiple OSes. I can look into this in more detail later.

Jens,
Does this proposal work for you? Also would like to here from you on three
big issues I mentioned earlier in this thread.

paul

On Mon, May 15, 2017 at 2:26 AM, Sitsofe Wheeler <notifications@github.com>
wrote:

> Bear in mind there's a chicken and an the egg issue here: because the
> commit is so big it takes review to find out what you don't like about it
> but it's hard to review because it's so big etc.
>
> Non-starters you didn't mention for me you didn't directly mention (i.e.
> things other than changes in defaults) in the current commit:
>
>    - now variable being changed to a long.
>    - nanosleep change which introduces a change but just comments out the
>    old code. Really needs a small reproducer to show the problem.
>
>
>    1.
>
>    Sitsofe mentioned that are quite a number of restrictions with this
>    new tracking log.
>
>    The restrictions that Paul refers to are mostly mentioned in
>    https://github.com/phoulihan9/fio/pull/1/files/
>    7aaeea5a483e59d356432a469bd897298a495192#diff-
>    05bc0511c8db43874bb452c70f7b6ddfR2626
>    <https://github.com/phoulihan9/fio/pull/1/files/7aaeea5a483e59d356432a469bd897298a495192#diff-05bc0511c8db43874bb452c70f7b6ddfR2626>
>    . For convenience I'll reproduce them here
>
>    +    back on verification described under the verify option. The restrictions when+    enabling the verify_track option are:+            - randommap is required+            - softrandommap is not supported+            - lfsr random generator not supported when using multiple block sizes+            - stonewall option required when more than one job present+            - file size must be an even multiple of the block size when iodepth > 1+            - verify_backlog not supported when iodepth > 1+            - verify_async is not supported+            - file sharing between concurrent jobs not supported+            - numjobs must be 1+            - io_submit_mode must be set to "inline"+            - verify=null or pattern are not supported+            - verify_only is not supported+            - io_submit_mode must be set to 'inline'+            - supplying a sequence number with rw option is not supported+            - experimental_verify is not supported
>
>    The ones that stand out are:
>    - stonewall required (even if the files aren't shared?)
>       - verify_backlog not supported when iodepth > 1
>       - numjobs being 1 (even if the files aren't shared?),
>       - verify_only is not supported (doesn't that make multiple
>       re-verifications harder?)
>
>    (PS: io_submit_mode seems to be listed twice)
>
>    It's worth noting that there are definitely fio users using bsrange
>    with verification because they report corner cases where it causes spurious
>    mismatches. Personally, I've recently been using fio (with patches!) to do
>    non-uniform verified overlapping I/O (in my case by using a
>    random_distribution).
>
> Re fio-tracking: forks happen and I'd vastly prefer a fork than this work
> being lost/unpublished :-) You wouldn't have to go so far as naming your
> fork entirely differently but perhaps tweak the VERSION string so it's
> clear you're not mainstream fio. However, there's a lot in this commit that
> I think can go into upstream fio so if a fork were to occur perhaps it
> would be possible to periodically submit small pieces of the fork upstream
> and thus shrinking it over time...
>
> (PS: @phoulihan9 <https://github.com/phoulihan9> your replies look a bit
> weird on github's web interface. You might find using github's web
> interface useful if you need to make a more complicated/intricate reply)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/phoulihan9/fio/pull/1#issuecomment-301386994>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AalS2KTV7yu1WmeuuX8jRoBTaUmfkxv5ks5r5_ApgaJpZM4NTYRy>
> .
>

sitsofe commented:
@phoulihan9 Something like this:
```
fio --iodepth=32 --verify=pattern --verify_dump=1 --verify_fatal=1 --size=100M --bsrange=512-4k --rw=randwrite --filename=/tmp/tmp.fio --verify_pattern="%o" --name=verifyoverlap
```

Should trigger partial overlaps but perhaps the checksum code can cope with this anyway? Regular fio needs something like https://github.com/sitsofe/fio/commit/4c8cdeb60f8352a000c052107571f76b02d913a1 to detect and cope with the overlap.

sitsofe commented:
@phoulihan9 More personally, I've had to use workloads like this
```
fio --randrepeat=0 --iodepth 32 --filename /dev/iscsidisk --rw randwrite --serialize_overlap 1 --time_based --runtime=1d --bsrange 512-4k --random_distribution=zipf:0.2 --exitall_on_error --name verify --verify crc32c --verify_backlog=500000 --verify_fatal=1 --verify_dump=1
```

To check a rapid overlap workload was performing correctly in problem scenarios.
